plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/tree/modifiers.kt

```diff
@@ -144,15 +146,36 @@ inline fun JKModifiersListOwner.forEachModifier(action: (JKModifierElement) -> U
 
 internal fun JKModifierElement.isRedundant(): Boolean {
     val owner = parent ?: return false
+    if (owner is JKMethod && owner.hasRedundantVisibility && modifier is Visibility) return true
+
     val hasOverrideModifier = (owner as? JKOtherModifiersOwner)?.hasOtherModifier(OVERRIDE) == true
+    val hasInnerModifier = (owner as? JKOtherModifiersOwner)?.hasOtherModifier(INNER) == true
     val isOpenAndAbstractByDefault = owner.let {
         (it is JKClass && it.isInterface()) ||
                 (it is JKDeclaration && it.parentOfType<JKClass>()?.isInterface() == true)
     }
+    val parentClass = owner.parentOfType<JKClass>()
+
+    val parentIsPrivate = parentClass?.visibility == PRIVATE || owner.parentOfType<JKMethod>() != null
+    val redundantOnConstructor = (owner is JKConstructor && parentClass?.classKind == JKClass.ClassKind.ENUM)
+            || (owner is JKKtPrimaryConstructor && parentClass?.visibility == PRIVATE && parentClass.parentOfType<JKClass>() != null)
 
     return when (modifier) {
-        PUBLIC, FINAL -> !hasOverrideModifier
+        PUBLIC -> !hasOverrideModifier || (owner is JKMethod && owner.isRedundantPublicBuiltIn())
+        FINAL -> !hasOverrideModifier
+        PRIVATE -> (parentIsPrivate && owner is JKMethod && owner !is JKConstructor) || redundantOnConstructor
+        INTERNAL -> (parentIsPrivate && owner is JKMethod && owner !is JKConstructor)
+                || redundantOnConstructor || (hasInnerModifier && parentIsPrivate)
         OPEN, ABSTRACT -> isOpenAndAbstractByDefault
+        PROTECTED -> owner is JKMethod && owner.name.value == "clone"
+                && owner.parameters.isEmpty() && owner.returnType.type.fqName == "kotlin.Any"
         else -> false
     }
-}
\ No newline at end of file
+}
+
+private fun JKMethod.isRedundantPublicBuiltIn() = when {
+    name.value == "toString" && parameters.isEmpty() && returnType.type.fqName == "kotlin.String" -> true
+    name.value == "equals" && parameters.size == 1 && returnType.type.fqName == "kotlin.Boolean" -> true
```

Comment:
I believe the parameter type should also be checked, otherwise it may not be an override. We should also add a test for this case.